drivers/can/can_rcar.c

```diff
@@ -0,0 +1,1042 @@
+	tx_cb->cb = callback;
+	tx_cb->cb_arg = callback_arg;
+
+	k_sem_reset(&tx_cb->sem);
+
+	data->tx_head++;
+
+	if (msg->id_type == CAN_STANDARD_IDENTIFIER) {
+		identifier = msg->id << RCAR_CAN_MB_SID_SHIFT;
+	} else {
+		identifier = msg->id | RCAR_CAN_MB_IDE;
+	}
+
+	if (msg->rtr == CAN_REMOTEREQUEST) {
+		identifier |= RCAR_CAN_MB_RTR;
+	}
+
+	sys_write32(identifier, config->reg_addr + RCAR_CAN_MB_56);
+
+	sys_write16(msg->dlc, config->reg_addr
+		    + RCAR_CAN_MB_56 + RCAR_CAN_MB_DLC_OFFSET);
+
+	for (i = 0; i < msg->dlc; i++) {
+		sys_write8(msg->data[i], config->reg_addr
+			   + RCAR_CAN_MB_56 + RCAR_CAN_MB_DATA_OFFSET + i);
+	}
+
+	/* Start Tx: increment the CPU-side pointer for the transmit FIFO
+	 * to the next mailbox location
+	 */
+	sys_write8(0xff, config->reg_addr + RCAR_CAN_TFPCR);
+
+	k_mutex_unlock(&data->inst_mutex);
+	if (callback == NULL) {
+		k_sem_take(&tx_cb->sem, K_FOREVER);
+	}
+
+	return CAN_TX_OK;
+}
+
+static inline int can_rcar_attach(const struct device *dev,
+				  can_rx_callback_t cb,
+				  void *cb_arg,
+				  const struct zcan_filter *filter)
+{
+	struct can_rcar_data *data = DEV_CAN_DATA(dev);
+	int i;
+
+	for (i = 0; i < CONFIG_CAN_RCAR_MAX_FILTER; i++)
+		if (data->rx_callback[i] == NULL) {
+			data->rx_callback[i] = cb;
+			data->rx_callback_arg[i] = cb_arg;
+			data->filter[i] = *filter;
+			return i;
+		}
+	return CAN_NO_FREE_FILTER;
+}
+
+int can_rcar_attach_isr(const struct device *dev, can_rx_callback_t isr,
+			void *cb_arg,
+			const struct zcan_filter *filter)
+{
+	struct can_rcar_data *data = DEV_CAN_DATA(dev);
+	int filter_nr;
+
+	k_mutex_lock(&data->rx_mutex, K_FOREVER);
+	filter_nr = can_rcar_attach(dev, isr, cb_arg, filter);
+	k_mutex_unlock(&data->rx_mutex);
+	return filter_nr;
+}
+
+void can_rcar_detach(const struct device *dev, int filter_nr)
+{
+	struct can_rcar_data *data = DEV_CAN_DATA(dev);
+
+	if (filter_nr >= CONFIG_CAN_RCAR_MAX_FILTER) {
+		return;
+	}
+
+	k_mutex_lock(&data->rx_mutex, K_FOREVER);
+	data->rx_callback[filter_nr] = NULL;
+	k_mutex_unlock(&data->rx_mutex);
+}
+
+static int can_rcar_init(const struct device *dev)
+{
+	const struct can_rcar_cfg *config = DEV_CAN_CFG(dev);
+	struct can_rcar_data *data = DEV_CAN_DATA(dev);
+	struct can_timing timing;
+	int ret;
+	uint16_t ctlr;
+	uint8_t idx;
+
+	k_mutex_init(&data->inst_mutex);
+	k_mutex_init(&data->rx_mutex);
+	k_sem_init(&data->int_sem, 0, 1);
+	k_sem_init(&data->tx_sem, RCAR_CAN_FIFO_DEPTH, RCAR_CAN_FIFO_DEPTH);
+	for (idx = 0; idx < RCAR_CAN_FIFO_DEPTH; idx++) {
+		k_sem_init(&data->tx_cb[idx].sem, 0, 1);
+	}
+
+	data->tx_head = 0;
+	data->tx_tail = 0;
+	data->int_thread_stack = can_rcar_thread_stack;
+	memset(data->rx_callback, 0, sizeof(data->rx_callback));
+	data->state = CAN_ERROR_ACTIVE;
+	data->state_change_isr = NULL;
+
+	/* reset the registers */
+	ret = clock_control_off(config->clock_dev,
+				(clock_control_subsys_t *)&config->mod_clk);
+	if (ret < 0) {
```

Comment:
if (ret != 0) for homogeneity ?

Reply:
+ 854 + 860 + 866 + 872 + 892 + 898 + 903 + 952

Reply:
return 0 on success, negative errno on failure
sounds good